sympy/functions/combinatorial/factorials.py

```diff
@@ -866,192 +941,455 @@ def fdiff(self, argindex=1):
         else:
-            return binomial(*self.args)
+            return self
+        i = Dummy('never_seen')
+        return (Product(n - k + i, (i, 1, k))/factorial(k)).doit()
+
+    def _eval_rewrite_as_Piecewise(self, n, k, **kwargs):
+        from sympy import gamma, Abs, Lt
+        isint = lambda x: Eq(x, floor(x))
+        nneg = lambda x: Or(Eq(x, 0), Eq(x/Abs(x), 1))
+        for i in igen():
+            if not n.has(i) and not k.has(i):
+                break
+        return Piecewise(
+            (factorial(n)/factorial(k)/factorial(n - k),
+                And(isint(n), isint(k),
+                nneg(n), nneg(k), nneg(n - k))),
+            (ff(n, k)/factorial(k),
+                And(isint(k), nneg(k))),
+            (ff(n, n - k)/factorial(n - k),
+                And(isint(n - k), nneg(n - k))),
+            (0, And(*[i for x in (n, k, n - k)
+                for i in [isint(x), Lt(x, 0)]])),
+            (gamma(n + 1)/gamma(k + 1)/gamma(n - k + 1),
+                True))
 
     def _eval_rewrite_as_factorial(self, n, k, **kwargs):
-        return factorial(n)/(factorial(k)*factorial(n - k))
+        from sympy import Pow
+        isint = lambda x: x.is_integer is True
+        nneg = lambda x: x.is_nonnegative is True
+        neg = lambda x: x.is_negative is True
+        if And(isint(n), isint(k)):
+            if And(neg(n), nneg(k)):
+                return(Pow(-1, k)*factorial(k - n - 1)/(factorial(k)*factorial(-n - 1)))
+            if And(neg(n), neg(k)):
+                return (Pow(-1, n - k)*factorial(-k - 1)/(
+                    factorial(n - k)*factorial(-n - 1)))
+            if And(nneg(n), neg(k)):
+                return S.Zero
+            if And(nneg(n), nneg(k)):
+                return factorial(n)/(factorial(k)*factorial(n - k))
 
     def _eval_rewrite_as_gamma(self, n, k, **kwargs):
-        from sympy import gamma
-        return gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1))
+        from sympy import gamma, Pow
+        ki = k.is_integer
+        nki = (n - k).is_integer
+        if ki is None:
+            return
+        if nki is None:
+            return
+        if n.is_integer and k.is_integer:
+            if n.is_negative:
+                if k.is_nonnegative:
+                    return(Pow(-1, k)*gamma(k - n)/(gamma(-n)*gamma(k + 1)))
+                if k.is_negative:
+                    return (Pow(-1, n - k)*gamma(-k)/(
+                        gamma(-n)*gamma(n - k + 1)))
+            elif n.is_nonnegative:
+                if k.is_negative:
+                    return S.Zero
+                if k.is_nonnegative:
+                    return (gamma(n + 1)/(gamma(k + 1)*gamma(n - k + 1)))
+        else:
+            ok = False
+            if ki is False and nki is False:
+                ok = True
+            elif ki and (k + 1).is_positive:
+                ok = True
+            elif nki and (n - k + 1).is_positive:
+                ok = True
+            if ok:
```

Comment:
```suggestion
        elif fuzzy_and([n.is_integer, k.is_integer]) is False:
```
cause all the integer variations are addressed above right ( all variations of positive and negative ) 😁
only non-integers get this choice. Looks Okay ?
if any `integer=None` it simply wont rewrite ?
and when `integer=False` the negatives don't matter anymore so rewrite's are good to go.
This'll hold for the right assumptions.

Reply:
if n = 5/2 and k = 7/2, n - k + 1 is 0 and we would be dividing by zoo...and we can't have that because of the cancellation issue.